// @darkobits/log@2.0.0-beta.16/dist/etc/config.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.DEFAULT_STREAM = void 0;

var _env2 = _interopRequireDefault(require("@darkobits/env"));

var _over = _interopRequireDefault(require("over-9000"));

var _env;

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

const DEFAULT_STREAM = process.stderr;
exports.DEFAULT_STREAM = DEFAULT_STREAM;
const DEFAULT_CONFIG = {
  level: (_env = (0, _env2.default)('LOG_LEVEL')) !== null && _env !== void 0 ? _env : 'info',
  stripIndent: true,
  levels: {
    silent: {
      level: -Infinity,
      label: ''
    },
    error: {
      level: 1000,
      label: 'ERR!'
    },
    warn: {
      level: 2000,
      label: 'WARN'
    },
    notice: {
      level: 3000,
      label: 'notice'
    },
    http: {
      level: 4000,
      label: 'http'
    },
    timing: {
      level: 5000,
      label: 'timing'
    },
    info: {
      level: 6000,
      label: 'info'
    },
    verbose: {
      level: 7000,
      label: 'verb'
    },
    debug: {
      level: 8000,
      label: 'debug'
    },
    silly: {
      level: _over.default,
      label: 'sill'
    }
  }
};
var _default = DEFAULT_CONFIG;
exports.default = _default;
