// @darkobits/log@2.0.0-beta.16/dist/etc/style.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
const DEFAULT_STYLE = {
  style: {
    timestamp: (token, chalk) => chalk.keyword('gray')(`[${token}]`),
    heading: (token, chalk) => chalk.keyword('ghostwhite')(token),
    prefix: (token, chalk) => chalk.keyword('fuchsia')(token)
  },
  levels: {
    error: {
      style: (token, chalk) => chalk.keyword('orangered').bgKeyword('black')(token)
    },
    warn: {
      style: (token, chalk) => chalk.keyword('black').bgKeyword('gold')(token)
    },
    notice: {
      style: (token, chalk) => chalk.keyword('orange').bgKeyword('black')(token)
    },
    http: {
      style: (token, chalk) => chalk.keyword('lightseagreen')(token)
    },
    timing: {
      style: (token, chalk) => chalk.keyword('dodgerblue')(token)
    },
    info: {
      style: (token, chalk) => chalk.keyword('mediumseagreen')(token)
    },
    verbose: {
      style: (token, chalk) => chalk.keyword('royalblue').bgKeyword('black')(token)
    },
    debug: {
      style: (token, chalk) => chalk.keyword('gray')(token)
    },
    silly: {
      style: (token, chalk) => chalk.keyword('black').bgKeyword('gainsboro')(token)
    }
  }
};
var _default = DEFAULT_STYLE;
exports.default = _default;
module.exports = exports.default;
